File: dlo_export/dlo.php
<?php

DEFINE('BRIGHTSPACE_PATH', 'https://dlo.mijnhva.nl');
DEFINE('SIGN_IN_PATH', 'https://login.hva.nl/');

require __DIR__.'/rest.php';
require __DIR__.'/cookies.php';


class DLO {
    private $_rest;
    private $_cookies;

    public function __construct() {
        $this->_rest = new Rest();
        $this->_cookies = new Cookies();
    }

    public function login($username, $password) {
        // First visit the page to initialize mostly the cookies
        $firstInit = $this->Rest()->get_request(BRIGHTSPACE_PATH);

        // Strip the action path to sign-in from this first visit
        preg_match('/action=\"\/(adfs\/(.*?))\"/', $firstInit, $formActionMatches);
        $formActionPath = $formActionMatches[1];

        // Do login POST request to mijnhva
        $loginMijnHva = $this->Rest()->post_request(SIGN_IN_PATH . $formActionPath, ['UserName' => $username, 'Password' => $password]);

        if(preg_match('/Incorrect user ID/', $loginMijnHva))
            throw new Exception("Username/password invalid");

        // Get the tokens from the login
        preg_match('/action=\"(.*?)\">/', $loginMijnHva, $surfconextFormMatches);
        preg_match('/name=\"SAMLResponse\" value=\"(.*?)\"/', $loginMijnHva, $SAMLResponseFormMatches);

        $surfconextFormAction = $surfconextFormMatches[1];
        $SAMLResponseForm = $SAMLResponseFormMatches[1];

        // Do some magic with surfconext and the tokens
        $surfconextAccess = $this->Rest()->post_request($surfconextFormAction, ['SAMLResponse' => $SAMLResponseForm]);

        preg_match('/action=\"(.*?)\">/', $surfconextAccess, $surfconextFormMatches);
        preg_match('/name=\"SAMLResponse\" value=\"(.*?)\"/', $surfconextAccess, $SAMLResponseFormMatches);
        preg_match('/name=\"RelayState\" value=\"(.*?)\"/', $surfconextAccess, $relayStateFormMatches);

        $dloSsoFormAction = $surfconextFormMatches[1];
        $dloSAMLResponseForm = $SAMLResponseFormMatches[1];
        $dloRelayStateForm = $relayStateFormMatches[1];

        // Login with SSO
        $surfconextAccess = $this->Rest()->post_request($dloSsoFormAction, ['SAMLResponse' => $dloSAMLResponseForm, 'RelayState' => $dloRelayStateForm]);
        // Remove whitespaces
        $surfconextAccessCleaned = preg_replace('/\s+/', ' ', $surfconextAccess);

        // Finally, login process, get last redirect url
        preg_match('/window.location = \'(.*?)\'/', $surfconextAccessCleaned, $processLoginActionUrlMatches);
        $processLoginActionUrl = $processLoginActionUrlMatches[1];

        $dloHomepage = $this->Rest()->get_request(BRIGHTSPACE_PATH . $processLoginActionUrl);

        return (bool)preg_match('/Session.UserId/', $dloHomepage);
    }

    public function getCourses($courseTypeCode = 'Course Offering') {
        $groupsPage = $this->Rest()->json($this->Rest()->get_request(BRIGHTSPACE_PATH . '/d2l/api/lp/1.10/enrollments/myenrollments/'));
        $allMyGroups = [];

        while(true) {
            $allMyGroups = $allMyGroups + $groupsPage['Items'];

            if(!empty($groupsPage['PagingInfo']['Bookmark']))
                $groupsPage = $this->Rest()->json($this->Rest()->get_request(BRIGHTSPACE_PATH . '/d2l/api/lp/1.10/enrollments/myenrollments/?bookmark=' . $groupsPage['PagingInfo']['Bookmark']));
            else
                break;
        }

        $returnGroups = [];
        foreach($allMyGroups as $index => $group)
            if($group['OrgUnit']['Type']['Code'] == $courseTypeCode)
                $returnGroups[] = $group;

        return $returnGroups;
    }

    public function getGroupsByCourseId($courseId) {
        /*$classListRequest = $this->Rest()->json($this->Rest()->get_request(BRIGHTSPACE_PATH . '/d2l/api/le/1.10/41216/classlist/'));
        $classList = [];
        foreach($classListRequest as $user)
            $classList[$user['Identifier']] = $user['FirstName'] . ' ' . $user['LastName'];*/


        $categories = $this->Rest()->json($this->Rest()->get_request(BRIGHTSPACE_PATH . '/d2l/api/lp/1.10/' . $courseId . '/groupcategories/'));
        $courseCategoryGroupData = [];

        foreach($categories as $category) {
            $categoryGroupList = [];
            $categoryGroups = $this->Rest()->json($this->Rest()->get_request(BRIGHTSPACE_PATH . '/d2l/api/lp/1.10/' . $courseId . '/groupcategories/' . $category['GroupCategoryId'] . '/groups/'));

            foreach($categoryGroups as $group) {

                $categoryGroupList[] = $group;
            }

            $courseCategoryGroupData[] = [
                'category' => $category,
                'groups' => $categoryGroupList
            ];
        }

        return $courseCategoryGroupData;
    }



    private function Rest() {
        return $this->_rest;
    }

    private function Cookies() {
        return $this->_cookies;
    }

}

File: dlo_export/dlo.py
import requests
import re
import json

class DLO:
    BRIGHTSPACE_PATH = "https://dlo.mijnhva.nl"
    SIGN_IN_PATH = "https://login.hva.nl/"
    session = requests.Session()

    def get_session(self):
        return self.session

    def make_request(self, path, type = 'get', data=None):
        if data is None:
            data = {}
        if type == 'get':
            request = self.get_session().get(path, data=data)
        elif type == 'post':
            request = self.get_session().post(path, data=data)
        else:
            print('Unknown type')
            return

        return request.text

        if request.status_code == 200:
            return request.text
        return False

    def login(self, username, password):
        rq = self.make_request(self.BRIGHTSPACE_PATH)

        if not rq:
            print("Error -> Cannot find homepage")
            return

        #print(rq)
        form_action = re.search("action=\"/(adfs/(.*))\"", rq)

        if not form_action:
            print("Error -> Cannot find form action url")
            return

        form_action_path = form_action.group(1)

        login_rq = self.make_request(self.SIGN_IN_PATH + form_action_path, 'post', {
            'UserName': username,
            'Password': password,
            'AuthMethod': 'FormsAuthentication'
        })

        surfconext_form_action = re.search("action=\"(.*?)\">", login_rq)
        SAMLResponse_form = re.search("name=\"SAMLResponse\" value=\"(.*?)\"", login_rq)

        if not SAMLResponse_form:
            return False

        login_rq_surfconext = self.make_request(surfconext_form_action.group(1), 'post', {
            'SAMLResponse': SAMLResponse_form.group(1)
        })

        dlo_sso_form_action = re.search("action=\"(.*?)\">", login_rq_surfconext)
        dlo_SAMLResponse_form = re.search("name=\"SAMLResponse\" value=\"(.*?)\"", login_rq_surfconext)
        dlo_RelayState_form = re.search("name=\"RelayState\" value=\"(.*?)\"", login_rq_surfconext)

        dlo_login = self.make_request(dlo_sso_form_action.group(1), 'post', {
            'SAMLResponse': dlo_SAMLResponse_form.group(1),
            'RelayState': dlo_RelayState_form.group(1)
        })

        dlo_sso_redirect = re.search("window.location='(.*?)'", re.sub(r'\s+', '', dlo_login))

        dlo_home = self.make_request(self.BRIGHTSPACE_PATH + dlo_sso_redirect.group(1))

        if re.search("Session.UserId", dlo_home):
            return True
        return False


    def get_my_courses(self, course_type_code = 'Course Offering'):
        groups_page = json.loads(self.make_request(self.BRIGHTSPACE_PATH + "/d2l/api/lp/1.10/enrollments/myenrollments/"))
        all_groups = groups_page['Items']

        while groups_page['PagingInfo']['Bookmark']:
            groups_page = json.loads(self.make_request(self.BRIGHTSPACE_PATH + "/d2l/api/lp/1.10/enrollments/myenrollments/?bookmark=" + groups_page['PagingInfo']['Bookmark']))
            if groups_page['Items']:
                all_groups = all_groups + groups_page['Items']

        my_courses = []
        for group in all_groups:
            if group['OrgUnit']['Type']['Code'] == course_type_code:
                my_courses.append(group)

        return my_courses

    def get_all_students_from_course(self, course_id = None):
        return json.loads(self.make_request(self.BRIGHTSPACE_PATH + "/d2l/api/le/1.10/" + str(course_id) + "/classlist/"))